integration/proxy_helpers_test.go

```diff
@@ -0,0 +1,483 @@
+			err = trustedCluster.CheckAndSetDefaults()
+			require.NoError(t, err)
+
+			options.trustedCluster = trustedCluster
+
+		}
+	}
+}
+
+func mustGetCurrentUser(t *testing.T) *user.User {
+	user, err := user.Current()
+	require.NoError(t, err)
+	return user
+}
+
+func mustRunPostgresQuery(t *testing.T, client *pgconn.PgConn) {
+	result, err := client.Exec(context.Background(), "select 1").ReadAll()
+	require.NoError(t, err)
+	require.Equal(t, []*pgconn.Result{postgres.TestQueryResponse}, result)
+}
+
+func mustClosePostgresClient(t *testing.T, client *pgconn.PgConn) {
+	err := client.Close(context.Background())
+	require.NoError(t, err)
+}
+
+func k8ClientConfig(serverAddr, sni string) clientcmdapi.Config {
+	const clusterName = "gke_project_europecentral2a_cluster1"
+	return clientcmdapi.Config{
+		Clusters: map[string]*clientcmdapi.Cluster{
+			clusterName: {
+				Server:                serverAddr,
+				InsecureSkipTLSVerify: true,
+				TLSServerName:         sni,
+			},
+		},
+		Contexts: map[string]*clientcmdapi.Context{
+			clusterName: {
+				Cluster:  clusterName,
+				AuthInfo: clusterName,
+			},
+		},
+		CurrentContext: clusterName,
+	}
+}
+
+func mkPodList() *v1.PodList {
+	return &v1.PodList{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "PodList",
+			APIVersion: "v1",
+		},
+		Items: []v1.Pod{
+			{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "firstcontainer-66b6c48dd-bqmwk",
+				},
+			},
+		},
+	}
+}
+
+func startKubeAPIMock(t *testing.T) *httptest.Server {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/apis/authorization.k8s.io/v1/selfsubjectaccessreviews", func(rw http.ResponseWriter, request *http.Request) {
+	})
+	mux.HandleFunc("/api/v1/namespaces/default/pods", func(rw http.ResponseWriter, request *http.Request) {
+		rw.Header().Add("Content-Type", "application/json")
+		err := json.NewEncoder(rw).Encode(mkPodList())
+		require.NoError(t, err)
+	})
+
+	svr := httptest.NewTLSServer(mux)
+	t.Cleanup(func() {
+		svr.Close()
+	})
+	return svr
+}
+
+func mustCreateKubeConfigFile(t *testing.T, config clientcmdapi.Config) string {
+	configPath := filepath.Join(t.TempDir(), "config.yaml")
+	err := clientcmd.WriteToFile(config, configPath)
+	require.NoError(t, err)
+	return configPath
+}
+
+func mustStartALPNLocalProxy(t *testing.T, addr string, protocol alpnproxy.Protocol) *alpnproxy.LocalProxy {
+	listener, err := net.Listen("tcp", ":0")
+	require.NoError(t, err)
+
+	address, err := utils.ParseAddr(addr)
+	require.NoError(t, err)
+	lp, err := alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
+		RemoteProxyAddr:    addr,
+		Protocol:           protocol,
+		InsecureSkipVerify: true,
+		Listener:           listener,
+		ParentContext:      context.Background(),
+		SNI:                address.Host(),
+	})
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		lp.Close()
+	})
+
+	startSyncC := make(chan struct{})
+	go func() {
+		close(startSyncC)
+		err := lp.Start(context.Background())
+		require.NoError(t, err)
+	}()
+	<-startSyncC
```

Comment:
looks like `startSyncC` is closed immediately in the goroutine, does waiting on the channel really give you any guarantees?

Reply:
Right, it doesn't give any guarantees that `lp.Start` listener is running.  
The startC chan was suppose to make sure that the `svr.Serve` goroutine was scheduled.